### PR TITLE
Use sync.Pool for SubGraph nodes

### PR DESCRIPTION
**Pull request by ashwin95r**
This change is [Reviewable](https://reviewable.io/reviews/dgraph-io/dgraph/153)

### CODE REVIEW

**Comment by manishrjain** [comment]
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[query/query.go, line 132 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/153#-KOmVjYFbOa2XgB32KWB:-KOmVjYGM3XsXTiFp-OJ:b-xtlrob) ([raw file](https://github.com/dgraph-io/dgraph/blob/15e9dcaef4bbc6f55668f38ebd12e152e5e5acf0/query/query.go#L132)):*
> ```Go
> }
> 
> func newSubgraphNode() *SubGraph {
> ```

Can you please take a look at @pawanrawal  corresponding change. It avoids some of these functions.

---

*[query/query.go, line 280 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/153#-KOmW0BFJvkeCUy2th6q:-KOmW0BFJvkeCUy2th6r:bx3iff0) ([raw file](https://github.com/dgraph-io/dgraph/blob/15e9dcaef4bbc6f55668f38ebd12e152e5e5acf0/query/query.go#L280)):*
> ```Go
> func (g *SubGraph) ToJson(l *Latency) (js []byte, rerr error) {
> 	r, err := postTraverse(g)
> 	go releaseSubgraph(g)
> ```

You shouldn't be doing this while you're calling a function of the SubGraph itself.

---

*[query/query.go, line 413 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/153#-KOmW80gZOuXFTE_gUfF:-KOmW80gZOuXFTE_gUfG:b-kdcyn8) ([raw file](https://github.com/dgraph-io/dgraph/blob/15e9dcaef4bbc6f55668f38ebd12e152e5e5acf0/query/query.go#L413)):*
> ```Go
> 	}
> 
> 	go releaseSubgraph(g)
> ```

Shouldn't do this from within the SubGraph function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/153)*

**Comment by ashwin95r** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[query/query.go, line 280 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/153#-KOmW0BFJvkeCUy2th6q:-KOmrztsA9ibRLhKZHup:b-mysu36) ([raw file](https://github.com/dgraph-io/dgraph/blob/15e9dcaef4bbc6f55668f38ebd12e152e5e5acf0/query/query.go#L280)):*
<details><summary><i>Previously, manishrjain (Manish R Jain) wrote…</i></summary>
> You shouldn't be doing this while you're calling a function of the SubGraph itself.
</details>
Will a `defer`be good? Given it's called after the function returns.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/153)*

**Comment by manishrjain** [comment]
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[query/query.go, line 280 \[r1\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/153#-KOmW0BFJvkeCUy2th6q:-KOnHeoohktjR0hTjtjv:b-7sjarz) ([raw file](https://github.com/dgraph-io/dgraph/blob/15e9dcaef4bbc6f55668f38ebd12e152e5e5acf0/query/query.go#L280)):*
<details><summary><i>Previously, ashwin95r (Ashwin Ramesh) wrote…</i></summary>
> Will a `defer`be good? Given it's called after the function returns.
</details>
It's still ugly, because someone else is holding a pointer that you're releasing. The owner of this pointer should be the one releasing it.

---

*[query/query.go, line 390 \[r2\]](https://reviewable.io:443/reviews/dgraph-io/dgraph/153#-KOnI69Y0lBFNDmoymsx:-KOnI69Y0lBFNDmoymsy:b6zxbqk) ([raw file](https://github.com/dgraph-io/dgraph/blob/34306b1f0cb74b0599978c5db8129f8a80e4093e/query/query.go#L390)):*
> ```Go
> // predicate-entity based protocol buffer subgraph.
> func (g *SubGraph) ToProtocolBuffer(l *Latency) (n *graph.Node, rerr error) {
> 	defer releaseSubgraph(g)
> ```

No. Remove this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dgraph-io/dgraph/153)*

File: query/query.go [hunk 1]
```diff
@@ -122,6 +123,24 @@ type SubGraph struct {
...
+	},
+}
+
+func releaseSubgraph(root *SubGraph) {
+	if root == nil {
+		return
+	}
+
+	for _, child := range root.Children {
+		go releaseSubgraph(child)
```
Comment (anacrolix) [inline query/query.go]: This will cost you more than the Pool will ever give back. Get rid of the `go` here and you should see a big improvement.

**Pull request closed without merging**